pkg/controller/volume/scheduling/scheduler_binder.go

```diff
@@ -783,3 +833,72 @@ func (a byPVCSize) Less(i, j int) bool {
 func claimToClaimKey(claim *v1.PersistentVolumeClaim) string {
 	return fmt.Sprintf("%s/%s", claim.Namespace, claim.Name)
 }
+
+// isCSIMigrationOnForPlugin checks if CSI migrartion is enabled for a given plugin.
+func isCSIMigrationOnForPlugin(pluginName string) bool {
+	switch pluginName {
+	case csiplugins.AWSEBSInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationAWS)
+	case csiplugins.GCEPDInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationGCE)
+	case csiplugins.AzureDiskInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationAzureDisk)
+	case csiplugins.CinderInTreePluginName:
+		return utilfeature.DefaultFeatureGate.Enabled(features.CSIMigrationOpenStack)
+	}
+	return false
+}
+
+// isPluginMigratedToCSIOnNode checks if an in-tree plugin has been migrated to a CSI driver on the node.
+func isPluginMigratedToCSIOnNode(pluginName string, csiNode *storagev1beta1.CSINode) bool {
```

Comment:
why do we care that it must be migrated on the node for this conversion. It looks like we're just trying to do a best effort conversion.

I guess my question is - if we continue to use the in-tree plugin to do attach/detach/mount/etc in the case of fallback (because node is on too low of a version), is it still valid to do the conversion attempted in this PR? If so, we don't need to check the node at all.

Reply:
FWIW right now `isPluginMigratedToCSIOnNode` type logic is only used for Attach/Detach because of https://github.com/kubernetes/community/blob/master/contributors/design-proposals/storage/csi-migration.md#adc-and-kubelet-csiin-tree-sync

Please give a quick read - would be happy to chat more about this. I'm not sure I understand the intention of this code enough to give a recommendation either way. My hunch is probably that we *don't* want to check the node though.

Reply:
Thanks for pointing the design doc. I get your point now, and it makes sense to remove this logic here. Any objections if remove it? (CC @msau42)

PS: There's also similar logic in the volume limits predicate: https://github.com/kubernetes/kubernetes/blob/master/pkg/scheduler/algorithm/predicates/utils.go#L132-L148. Should it be dropped there as well?

Reply:
yes, I think we should drop it in both places

Reply:
For reference, the decision to implement this logic the CSI predicate was made based on this thread: https://github.com/kubernetes/enhancements/pull/942#discussion_r277060459

Reply:
I think we do need to keep the check if migration is enabled on the node.  Think about the scenario where nodes have not installed the CSI driver yet.  Then the Node will not have the new CSI topology labels on it.  So if we translate it here, then the scheduler predicate will fail because the new CSI label doesn't exist.

I think similar issue exists for attach limits?  If the csi driver is not installed, then the CSINode won't have the resource.